### PR TITLE
Integrate astropy-specific asdf tags

### LINKED ISSUE #6560: Consider moving astropy-specific schemas and tags from ASDF to astropy core
ASDF is a data format that is capable of serializing arbitrary Python objects. While any ASDF file can be reconstructed in memory using basic Python types (strings, lists, dicts), it is often useful to recreate the original object. Currently, ASDF maintains schemas (data type definitions) and tags (implementations of serialization) for various astropy objects. It would be useful to move the astropy-specific schemas and tags to astropy, where they can be maintained and testing in parallel with the corresponding objects.

@eteq

### PR DESCRIPTION
**Pull request by drdavella**
As discussed during the astropy coordination meeting, we would like to move astropy-specific tag implementations from ASDF to the astropy core package. Moving the tags to astropy will allow them to be maintained and tested in conjunction with any changes to the types they represent. This PR represents the first step in this direction.

Tags are classes that implement serialization/deserialization of specific types to/from YAML. These tags have been moved from ASDF with only very minor modifications to support their inclusion in astropy. As a first pass, I don't think it's worth reviewing the code under `astropy/io/asdf/tags` in any great detail since this code has been moved largely as-is. The remaining code is infrastructure that allows these tag classes to be discovered and used by the ASDF module.

No schemas have been moved from ASDF to astropy. Eventually I would like to move at least one or two schemas (perhaps including the one for FITS), but this can be done in a separate PR.

This PR does **not** introduce a hard dependency on `asdf`. Currently, one test in travis has been configured to install `asdf` from a branch on github. However, all other test runs should continue to pass without having `asdf` installed.

Merging this PR requires coordination with a corresponding PR in ASDF: https://github.com/spacetelescope/asdf/pull/359

As @taldcroft and I have discussed, if this PR is accepted, eventually it would be useful to try to combine ECSV code for serialization with the code in these tags since they serve very similar purposes.

EDIT: Fix #6560

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -103,7 +105,7 @@ matrix:
         - os: linux
           env: PYTHON_VERSION=3.5 SETUP_CMD='test --remote-data=astropy -a "--mpl"'
                CONDA_DEPENDENCIES=$CONDA_ALL_DEPENDENCIES
-               PIP_DEPENDENCIES='jplephem pytest-mpl'
+               PIP_DEPENDENCIES="jplephem pytest-mpl $ASDF_GIT"
```
Comment (drdavella) [inline .travis.yml]: Maybe I should add this dependency to the test on Py3.6 as well? Or maybe just that one instead of this one?

File: astropy/io/asdf/extension.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from astropy.io.asdf.tags.transform.projections import *
+from astropy.io.asdf.tags.transform.tabular import *
+from astropy.io.asdf.tags.unit.quantity import *
+from astropy.io.asdf.tags.unit.unit import *
+from astropy.io.asdf.types import _astropy_asdf_types
+
+from asdf.resolver import Resolver, DEFAULT_URL_MAPPING
+
+
+SCHEMA_PATH = os.path.abspath(
```
Comment (drdavella) [inline astropy/io/asdf/extension.py]: This is currently not used anywhere.

Comment (pllim) [reply astropy/io/asdf/extension.py]: Is there a reason to keep it around then?

**Comment by pllim** [comment]
Is it too early to ask about unified I/O with `tables` (or is that even possible)?

File: astropy/io/asdf/extension.py (hunk 2 above)
Comment (drdavella) [reply astropy/io/asdf/extension.py]: Not right now. Eventually if/when schemas are added this will be necessary.

Comment (pllim) [reply astropy/io/asdf/extension.py]: Perhaps comment it out until it is used?

File: .travis.yml (hunk 1 above)
Comment (bsipocz) [reply .travis.yml]: either case sounds good to me.

Comment (drdavella) [reply .travis.yml]: I realized it at least has to be in the coverage test or else the coverage metrics get all messed up. I think I'll leave it in both, though.

**Comment by pllim** [comment]
Does this partly or fully address #6560?

**Comment by drdavella** [comment]
@pllim yes it closes #6560, although if this PR is accepted we may decide to move a few schemas here as well.

File: astropy/io/asdf/extension.py (hunk 2 above)
Comment (mhvk) [reply astropy/io/asdf/extension.py]: I'd say remove it if it is not needed in this PR.

File: astropy/io/asdf/setup_package.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+                schemas.append(
+                    os.path.relpath(os.path.join(node, fname), root))
+
+    # In the package directory, install to the subdirectory 'schemas'
+    schemas = [os.path.join('schemas', s) for s in schemas]
+
+    return {'astropy.io.asdf': schemas}
+
+
+def requires_2to3():
```
Comment (mhvk) [inline astropy/io/asdf/setup_package.py]: Why is this needed?

File: astropy/io/asdf/extension.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# -*- coding: utf-8 -*-
+
+import os
+
+# Make sure that all tag implementations are imported by the time we create
```
Comment (mhvk) [inline astropy/io/asdf/extension.py]: Is this something that can be changed later? I'd assume it would be done in `tags.__init__` - is it not possible to have a `try/except` around an `asdf` import there?

I guess I'm mostly asking to clarify the comment here even more, including what the future final state should be.

File: astropy/io/asdf/extension.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+
+SCHEMA_PATH = os.path.abspath(
+    os.path.join(os.path.dirname(__file__), 'schemas'))
+
+
+class AstropyExtension(object):
+    @property
+    def types(self):
+        return _astropy_asdf_types
```
Comment (nden) [inline astropy/io/asdf/extension.py]: I may be missing something but it looks like `_astropy_asdf_types` includes only types with schemas defined in asdf. Are you going to create a different base type class for tags with schemas in astropy? If so, just ignore this, I will review this later.
(I am judging by code and comments in asdf/types.py.

File: astropy/io/asdf/extension.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    @property
+    def tag_mapping(self):
+        return [('tag:stsci.edu:asdf',
+                 'http://stsci.edu/schemas/asdf{tag_suffix}')]
+
+    @property
+    def url_mapping(self):
+        return DEFAULT_URL_MAPPING
```
Comment (nden) [inline astropy/io/asdf/extension.py]: Hm, I think `DEFAULT_URL_MAPPING` will work with schemas defined in asdf but not with schemas defined in astropy. Do we need another extension for schemas in astropy?

Comment (nden) [reply astropy/io/asdf/extension.py]: It looks like this class can be rewritten as:

```
class AstropyExtension(BuiltinExtension):
    @property
    def types(self):
        return _astropy_asdf_types
```

And then perhaps another extension which maps also the schemas from astropy.

File: astropy/io/asdf/extension.py (hunk 5 above)
Comment (drdavella) [reply astropy/io/asdf/extension.py]: Yes, you are correct. If/when any schemas are provided by astropy itself, we will require a different extension for those types, along with a separate type registry. However, in the meantime it might be a good idea to name this `AstropyAsdfExtension`.

File: astropy/io/asdf/setup_package.py (hunk 3 above)
Comment (drdavella) [reply astropy/io/asdf/setup_package.py]: It's not; I think it was just a copy-paste error :)

File: astropy/io/asdf/extension.py (hunk 4 above)
Comment (drdavella) [reply astropy/io/asdf/extension.py]: It might be possible to do this with `try/except` but I think it would have to be done with every `__init__.py` in the `astropy/io/asdf` tree in order to avoid import errors at each level. I'm not sure if that would be any cleaner, but I'm open to suggestions.

File: astropy/io/asdf/extension.py (hunk 6 above)
Comment (drdavella) [reply astropy/io/asdf/extension.py]: See above, but yes we will need another extension if/when we add any schemas to astropy itself. And that's a good idea to just subclass ASDF's own `BuiltinExtension` in this case.

File: astropy/io/asdf/extension.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+from astropy.io.asdf.types import _astropy_asdf_types
+
+from asdf.resolver import Resolver, DEFAULT_URL_MAPPING
+
+
+SCHEMA_PATH = os.path.abspath(
+    os.path.join(os.path.dirname(__file__), 'schemas'))
+
+
+class AstropyExtension(object):
```
[the remark this answers is not in the document]
Comment (nden) [reply astropy/io/asdf/extension.py]: :+1: Got it. I need to change the extension in jwst then.

File: astropy/io/asdf/setup_package.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+# Licensed under a 3-clause BSD style license
+import os
+
+def get_package_data():
```
Comment (taldcroft) [inline astropy/io/asdf/setup_package.py]: Right now it appears this is doing nothing since there is no `schemas` dir.  Is that right?  If so are you quite sure some schemas will be included in 3.0, or perhaps should this file be left out until that subsequent PR?

**Comment by taldcroft** [comment]
I've been struggling a bit to decide on documentation requirements for this subpackage.  At one level it is mostly a helper for `asdf` and doesn't directly bring much to `astropy` itself.  Just the same it would be odd for an entire astropy subpackage to not have any docs or API reference.

I think there should be *some* kind of docs that at least tell people what this is, but I'm slightly hesitant to add a new line (`ASDF tag definitions (astropy.io.asdf)`) to the top-level docs index.  Thoughts @mhvk or @pllim ?

**Comment by taldcroft** [comment]
Apart from my two comments, this looks fine to me.

**Comment by mhvk** [comment]
If it isn't going to be stuff that's directly used, maybe it should go in `io.misc`? It does seem a bit like `io.misc.hdf5` right now, and similar to the `yaml` stuff already there too. An advantage of this is that the `io.misc` already explicitly mentions it is not for direct use, so it leaves us free to later move to a separate sub-module if we get `asdf.open` etc.

**Comment by taldcroft** [comment]
> If it isn't going to be stuff that's directly used, maybe it should go in io.misc?

Yes!  I kept looking at the `io.misc` line in the docs summary and thinking it would fit right in, but somehow didn't get to the next thought of "let's move it there".

**Comment by pllim** [comment]
Ops. I created an `io.asdf` label too early, it seems. :sweat_smile:

File: astropy/io/asdf/setup_package.py (hunk 8 above)
Comment (drdavella) [reply astropy/io/asdf/setup_package.py]: Yes, that's true. I went back and forth on this and decided to leave it in since it will likely be required eventually, but it wouldn't be a big deal to remove it if you thought that was best.

**Comment by drdavella** [comment]
@mhvk, @taldcroft should I go ahead and move it then? It seems reasonable to me. However, I've already had discussions with @astrofrog and @pllim about adding unified I/O support, so I'd like to consider whether it's just going to move right back to its own package in the near future.

**Comment by pllim** [comment]
> whether it's just going to move right back to its own subpackage in the near future

IMHO we should decide now and just make it permanent. Moving codes around introduce deprecation warnings and not very fun for some users (though I am guilty of moving things...). Even if it is in `io.misc` with HDF5, unified I/O is still possible, right?

**Comment by mhvk** [comment]
@drdavella - to me, it mostly depends on whether the module is going to be useful on its own - i.e., whether the goal is that it becomes similar to the other entries directly below `io` (`fits`, `ascii`, `votable`), where one can use the code directly and all of it is in astropy, or whether it remains similar to `hdf5` and `yaml`, where  code is accessed mostly indirectly, via `Table` (or perhaps `NDData`) and the unified I/O, and where it is mostly glue to help use a different package (that one has to have installed).

For your current PR, and even including the schemas, the `hdf5`/`yaml` model seems the most appropriate, so I think `io.misc` makes most sense. But the main question may be whether it is intended to merge all of `asdf` into astropy core. If so, then obviously it should be `io.asdf`.

For the in-between case, where `asdf` is meant to be merged eventually, but not before 3.0, it depends mostly on what "eventually" is, but my overall tendency would be to start with `io.misc`, so that every step of the way things make sense (for the same reason, I'd remove that `setup_package` if it is not currently useful!).

Anyway, bottom line questions: is `asdf` intended to be merged into astropy core? And, if so, on what timescale?

**Comment by drdavella** [comment]
@mhvk that seems reasonable. There is no plan to merge all of `asdf` into astropy. In fact, we are trying to make the `asdf` implementation entirely independent of astropy, which is part of the motivation for moving these astropy-specific tags here. So I will go ahead and move this into the `io.misc` subpackage.

**Comment by drdavella** [comment]
Current travis failure appears to be a transient URL timeout.

**Comment by taldcroft** [comment]
@drdavella - given the move, you should add some docs to the `io.misc` section which gives some context for what the `io.misc.asdf` subpackage is doing in Astropy and provides API docs for any public functions (modeled after `astropy.io.misc.hdf5` there).  This should include anything that is called by `asdf` itself.

You will also want to rename the whole section `Miscellaneous: HDF5, YAML, ASDF, pickle (astropy.io.misc)`

**Comment by drdavella** [comment]
@taldcroft I have added documentation. Right now `asdf` automatically identifies and loads `astropy` tags if they are installed, so there's not really anything in the way of public functions or classes that need to be mentioned in the documentation. 

This will probably change in the future, but it may be problematic since referencing classes and functions from this subpackage in sphinx documentation will require them to be imported. This will not be possible if `asdf` is not installed. Are there other subpackages that have similar issues?

**Comment by drdavella** [comment]
@taldcroft, @mhvk, @astrofrog: if/when review on this PR is accepted, will you let me actually hit the button to merge? I realize this is not typical, but this merge needs to be carefully coordinated with ASDF and with a few other repositories, so it would be useful if I could control the timing myself.

**Comment by drdavella** [comment]
@taldcroft, thanks! Just a heads up but right before I merge I'll need to update the CI scripts to point to the ASDF's master branch instead of the forked branch that I'm using right now.

**Comment by mhvk** [comment]
@drdavella - if it is easier, just keep your branch around a bit longer and make a new PR that changes the CI.

File: CHANGES.rst [hunk 9]
```diff
@@ -25,6 +25,11 @@ astropy.extern
 astropy.io.ascii
 ^^^^^^^^^^^^^^^^
 
+astropy.io.asdf
```
Comment (mhvk) [inline CHANGES.rst]: Note, this is now wrong! SHould be `astropy.io.misc.asdf`!

File: astropy/io/misc/asdf/tags/fits/fits.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# -*- coding: utf-8 -*-
+
+import numpy as np
+
+from astropy.io import fits
+from astropy.io.misc.asdf.types import AstropyAsdfType
```
Comment (mhvk) [inline astropy/io/misc/asdf/tags/fits/fits.py]: This may be more logical as a relative import: `from ... import AstropyAsdfType`

Also the `asdf` import should be moved up, below numpy, since it is not local.

File: astropy/io/misc/asdf/tags/fits/fits.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+    @classmethod
+    def reserve_blocks(cls, data, ctx):
+        for hdu in data:
+            if hdu.data is not None:
+                yield ctx.blocks.find_or_create_block_for_array(hdu.data, ctx)
+
+    @classmethod
+    def assert_equal(cls, old, new):
+        from numpy.testing import assert_array_equal
```
Comment (mhvk) [inline astropy/io/misc/asdf/tags/fits/fits.py]: Why not import this at the top?

File: astropy/io/misc/asdf/tags/fits/fits.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+                hdu = fits.BinTableHDU(data=data, header=header)
+            else:
+                hdu = fits.ImageHDU(data=data, header=header)
+            hdus.append(hdu)
+        hdulist = fits.HDUList(hdus)
+        return hdulist
+
+    @classmethod
+    def to_tree(cls, hdulist, ctx):
+        from astropy import table
```
Comment (mhvk) [inline astropy/io/misc/asdf/tags/fits/fits.py]: Why not import at the top?

File: astropy/io/misc/asdf/tags/fits/tests/test_fits.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# -*- coding: utf-8 -*-
+
+import os
+
+import pytest
+import numpy as np
+
+from astropy.io import fits
```
Comment (mhvk) [inline astropy/io/misc/asdf/tags/fits/tests/test_fits.py]: Should go after `asdf` import.

File: astropy/io/misc/asdf/tags/table/table.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# -*- coding: utf-8 -*-
+import numpy as np
+
+from astropy import table
+from astropy.io.misc.asdf.types import AstropyAsdfType
```
Comment (mhvk) [inline astropy/io/misc/asdf/tags/table/table.py]: Same comments about local import, and import ordering.

File: astropy/io/misc/asdf/tags/time/time.py [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+                if not isinstance(location[comp], Quantity):
+                    location[comp] = Quantity(location[comp], unit=unit)
+            location = EarthLocation.from_geocentric(
+                location['x'], location['y'], location['z'])
+
+        return time.Time(value, format=format, scale=scale, location=location)
+
+    @classmethod
+    def assert_equal(cls, old, new):
+        from astropy.coordinates import EarthLocation
```
Comment (mhvk) [inline astropy/io/misc/asdf/tags/time/time.py]: No reason not to import at the top

File: CHANGES.rst (hunk 9 above)
Comment (drdavella) [reply CHANGES.rst]: Good catch!

**Comment by mhvk** [comment]
@drdavella - if it is much easier to address my nitpicky comments in a follow-up PR, that is fine too!

File: astropy/io/misc/asdf/tags/fits/fits.py (hunk 12 above)
Comment (drdavella) [reply astropy/io/misc/asdf/tags/fits/fits.py]: When this was in asdf it made sense to lazily load imports from astropy like this since they were not guaranteed to be used. However, I don't think that makes sense now that this lives in astropy, and I don't think it is used consistently within the tags anyway, so I'll fix it up.

File: astropy/io/misc/asdf/tags/fits/fits.py (hunk 10 above)
Comment (mhvk) [reply astropy/io/misc/asdf/tags/fits/fits.py]: I have no strong feelings about this, though have a slight preference for sticking to the convention we have, i.e., import relatively. As I wrote above, this makes sense especially within `io.misc.asdf` - it means we could easily move it later to `io.asdf` if that became useful (this also suggests keeping things like `astropy.io.fits` as they are...

Comment (drdavella) [reply astropy/io/misc/asdf/tags/fits/fits.py]: Sounds like a reasonable compromise. I'll post an update shortly.

File: CHANGES.rst [hunk 16]
```diff
@@ -25,6 +25,11 @@ astropy.extern
 astropy.io.ascii
 ^^^^^^^^^^^^^^^^
 
+astropy.io..misc.asdf
```
Comment (bsipocz) [inline CHANGES.rst]: please move this under the already existing astropy.io.misc section. Probably won't need a separate subsection in the changelog for now as that's a low traffic one anyway.

File: docs/io/misc.rst [hunk 17]
```diff
@@ -1,6 +1,6 @@
-*****************************************************
-Miscellaneous: HDF5, YAML, pickle (`astropy.io.misc`)
-*****************************************************
+***********************************************************
+Miscellaneous: HDF5, YAML, ASDF, pickle (`astropy.io.misc`)
```
Comment (bsipocz) [inline docs/io/misc.rst]: Could you add ``asdf`` to the dependency lists, too? It's scattered all over in pip and rtd requirement files, and narrative docs.

Comment (drdavella) [reply docs/io/misc.rst]: It's not a required dependency so I'm not sure where it belongs. Should it just be added to `pip-requirements-dev` and `pip-requirements-doc`?

Comment (bsipocz) [reply docs/io/misc.rst]: I assume it's the exact same situation as for h5py, the module does some imports of asdf making it an optional dependency overall?

Comment (drdavella) [reply docs/io/misc.rst]: Can I do this in a separate PR? Or at least just leave a comment in the requirement file? I'm not sure this makes sense until the corresponding version of `asdf` is released.

Comment (bsipocz) [reply docs/io/misc.rst]: Sure. I totally missed the conversation above when I left these minor comments, so absolutely address these in a follow-up.

**Comment by drdavella** [comment]
@mhvk, @bsipocz I think I've covered all of the comments from your reviews in case you want to take another look.

**Comment by Cadair** [comment]
🎉

**Comment by bsipocz** [comment]
@drdavella - Since it has the label please add whatsnew text to http://docs.astropy.org/en/latest/whatsnew/3.0.html#what-s-new-in-astropy-3-0
